### ISSUE TITLE
Wrong configuration loaded if path is not `pyproject.toml`

### ISSUE THREAD

**Issue by lyz-code** [dbatten5/maison#141]
Hi @dbatten5, I'm facing an issue when loading the configuration from a different file than `pyproject.toml`. Imagine that I have three files with the same content:

```
.
├── config
│   └── pyproject.toml
├── other.toml
└── pyproject.toml
```

When loading them:

```python
>>> ProjectConfig(project_name="autoimport", source_files=['pyproject.toml']).to_dict()
{'common_statements': {'FooBar': 'from baz.qux import FooBar'}}
>>> ProjectConfig(project_name="autoimport", source_files=['other.toml']).to_dict()
{'tool': {'autoimport': {'common_statements': {'FooBar': 'from baz.qux import FooBar'}}}}
>>> ProjectConfig(project_name="autoimport", source_files=['config/pyproject.toml']).to_dict()
{'tool': {'autoimport': {'common_statements': {'FooBar': 'from baz.qux import FooBar'}}}}
```

I'd expect that loading `other.toml` and `config/pyproject.toml` returns the same result as `pyproject.toml` == `{'common_statements': {'FooBar': 'from baz.qux import FooBar'}}`

I'm using version `1.4.0`.

Thanks!

**Comment by dbatten5** [comment]
Hi @lyz-code good to hear from you again 👋 

I definitely agree that `config/pyproject.toml` should have the same result as `pyproject.toml`, that's a bug.

As for `other.toml`, I think that's up for discussion. I envisaged supporting generic `toml` files in situations where there's a dedicated `toml` file for the tool being used, eg. `autoimport.toml`. In that case I thought it would be odd if `autoimport.toml` had a structure like
```toml
[tool.autoimport]
common_statements = { "FooBar" = "from baz.qux import FooBar" }
```

Since `autoimport.toml` is dedicated to `autoimport` wouldn't it just list the config at the root level? Eg.
```toml
common_statements = { "FooBar" = "from baz.qux import FooBar" }
```

**Comment by lyz-code** [comment]
Yeah, what you say makes total sense

**Comment by dbatten5** [comment]
Hi @lyz-code sorry I haven't had a chance to look at this, I've been away for a while. Thinking about this again, I can't really envisage a situation where the `pyproject.toml` wouldn't be at the root of the project directory. Do you have any examples of this happening?

**Comment by lyz-code** [comment]
Hi @dbatten5 it's nice to hear from you again :), I had to dive deep into the code to remember why I needed this myself xD

It looks I was trying to [make a test to check the `--config-file` argument of `autoimport](https://github.com/lyz-code/autoimport/blob/main/tests/e2e/test_cli.py#L144). The idea was that the user could select a different `pyproject.toml` than the original. I myself don't have that need but some user made a PR to add that functionality, so I guess it needs it `¯\_(ツ)_/¯`. 

If it's hard to implement or it adds complexity, I can add a `DeprecationWarning` and if no one complains in 6 months we can close the issue

**Comment by dbatten5** [comment]
Hi @lyz-code, makes sense that a user would like to specify a different `autoimport` config file, but I would be surprised if they were trying to specify a `pyproject.toml` file at a level different to the project root as AFAIK the `pyproject.toml` has to be at the root of the project directory (couldn't really find much information on this so I could well be wrong about that). If they were trying to specify eg. a `autoimport.toml` at a different level to the root, maybe at `config/autoimport.toml` then this package would work as is right?

**Comment by dbatten5** [comment]
Actually, just taking a look at the code, I think if you added a `starting_path` argument to the constructor for `ProjectConfig` then you might be able to achieve what you're after
```python
ProjectConfig(project_name="autoimport", source_files=['pyproject.toml'], starting_path=Path("path/to/config").to_dict()
```

**Comment by lyz-code** [comment]
It makes sense all that you said. Whenever I can spare some time I'll try to implement your proposal

**Comment by marcules** [comment]
Hi @dbatten5 I ran into the same issue here as well - it looks like if you provide the path to the `pyproject.toml` directly, the simple check `if source.startswith("pyproject"):` is failing, because that's the start of the filename not the start of the (absolute) path.

https://github.com/dbatten5/maison/blob/main/src/maison/utils.py#L101

I would propose to change this line to `if file_path.name.startswith("pyproject"):`
So the path library can handle it and you would just check the basename instead.

**Comment by dbatten5** [comment]
Hi @marcules, ah yes good catch, will get that fixed up ASAP

**Comment by dbatten5** [comment]
Hey @lyz-code - slightly off topic but I'm currently experience some clashing dependencies with `flake8` and `mkdocs` which I think has been resolved by `flake8` releasing version `6.0.0` recently, however this version requires python version `>=3.8.1`. I'm happy to drop python `3.7` support in `maison` to fix that clash but would that be a pain for `autoimport`?

**Comment by lyz-code** [comment]
OMG `6.0.0` already? They did a big fuck up between `4.x` and `5.x` (I'm strugging with it and `flakeheaven` is too). For example `flakeheaven` does not yet support `5.x`. I'm sad to say that if you pin to something like `flake8>=5.0.0` I won't be able to use the newer versions until `flakeheaven` supports it :(. Although if it's in the dev dependencies, it won't affect me. 

My plan is to support python `3.7` until it's end of life on the 27 Jun 2023 (can't wait, if I'm honest with you), so I'd be unable to use newer versions if you also pin the python version.

That being said, go ahead :) , I've been struggling myself with dependencies clash these months (specially due to `flake8`) and I completely understand you don't want to fight with it anymore. If I feel that I need a new feature of `maison` more than the will to give support till the end of life or to use `flakeheaven` I'll upgrade myself.

On the other hand, you probably only need the `flake8` dependencies for the development environment. So maybe you can keep the minimum python version, the pins of your direct dependencies and then generate the `Poetry.lock` with the constrain of `python >=3.8`?

Thanks for asking :)

**Comment by dbatten5** [comment]
Hi @marcules sorry just coming back to this now. Could I confirm that you tried adding the `starting_path` argument to `ProjectConfig`? That should allow you to specify a `pyproject.toml` in a different location, e.g.:
```python
config = ProjectConfig(
  project_name="yamlfix",
  starting_path=Path("/some/other/path")
)

print(config.config_path)
#> PosixPath(/some/other/path/pyproject.toml)
```

**Comment by dbatten5** [comment]
@lyz-code yeah that makes sense. I think I wouldn't be able to upgrade `flake8` to `6.0.0` without updating the minimum python version to `3.8.1` even if `flake8` is just a dev dependency (using `poetry`)? Unless I'm missing something

**Comment by lyz-code** [comment]
[Then use `pdm`!!](https://lyz-code.github.io/blue-book/pdm/#solve-the-locking-failure)

But surely there's a way to tell `poetry` to not to resolve the constrains of some dependency and override it with the desired version

**Comment by dbatten5** [comment]
@lyz-code Ah yeah I've come across `pdm` - looks like a good tool althogh feels a bit of a pain to migrate. I've been looking at replacing `flake8` with `ruff` anyway so maybe I'll give that a try

**Comment by dbatten5** [comment]
closing due to inactivity as but feel free to reopen if there are still issues
